### ISSUE TITLE
Docker + pubsub + subprocess hangs with no errors

### ISSUE THREAD

**Issue by jtressle** [googleapis/python-pubsub#112]
I'm running pubsub in a docker container and am experiencing hangs after a few minutes.  I've been running pubsub with similar commands outside of docker, and see no hangs (over 1 year running).

I was able to reproduce my issue with a simplified implementation outlined below.

The following four files (docker file, requirements.txt, worker.py, test.py) are placed into the same directory.
[Dockerfile.txt](https://github.com/googleapis/python-pubsub/files/4732822/Dockerfile.txt)
[requirements.txt](https://github.com/googleapis/python-pubsub/files/4732823/requirements.txt)
[worker.py.txt](https://github.com/googleapis/python-pubsub/files/4732825/worker.py.txt)
[test.py.txt](https://github.com/googleapis/python-pubsub/files/4732827/test.py.txt)

To build the docker file, I ran 
`docker build --tag=test .`

To run the docker file, I did the following:

```
docker run -it \
   -e GOOGLE_APPLICATION_CREDENTIALS=/keys/service-account.json \
   -e GRPC_ENABLE_FORK_SUPPORT=false \
   -e GRPC_VERBOSITY=debug \
   -e GRPC_TRACE=all \
   -v /tmp/keys/:/keys/:ro \
   -v /tmp/logs:/logs \
test:latest bash
```

To simplify my issue, I used `subprocess.call_back()` to run test.py, which is basically a 3 second timer.  

I'm running docker on my Mac, but I've also experienced the same hanging using kubernetes, and on an Ubuntu 18.04 machine running docker.  The issue is that after a few minutes, the pubsub subscriber becomes unresponsive.  The above code runs perfectly in a virtual environment without docker.  

The following log file is the publisher output, and has the message ID's.  
[publish_output.txt](https://github.com/googleapis/python-pubsub/files/4732851/publish_output.txt)

The following log file is the subscriber output.  
[subscriber_output.txt](https://github.com/googleapis/python-pubsub/files/4732854/subscriber_output.txt)

The last successful message received was message ID 1253690400552289 (line 9338 of subscriber_output.txt).  Messages 1253713007710668, 1253714609061509, 1253713464329969 and onward were all not received by pubsub.

Is there any additional setting I need to do?  Unfortunately, I'm running custom C++ code from python so I need to use subprocesses.

Inside of docker, I ran `python worker.py` to start pubsub.

Any help is appreciated.

Thanks!

**Updated to show current configuration**

Below are my python packages pip downloaded:

```
CacheControl             0.12.6
cachetools               4.1.0
certifi                  2020.4.5.1
chardet                  3.0.4
firebase                 3.0.1
firebase-admin           4.3.0
firestore                0.0.8
google-api-core          1.18.0
google-api-python-client 1.9.1
google-auth              1.16.1
google-auth-httplib2     0.0.3
google-cloud-core        1.3.0
google-cloud-firestore   1.7.0
google-cloud-logging     1.15.0
google-cloud-pubsub      1.5.0
google-cloud-storage     1.28.1
google-resumable-media   0.5.1
googleapis-common-protos 1.52.0
grpc-google-iam-v1       0.12.3
grpcio                   1.29.0
httplib2                 0.18.1
idna                     2.9
iso8601                  0.1.12
msgpack                  1.0.0
pip                      20.1.1
protobuf                 3.12.2
pyasn1                   0.4.8
pyasn1-modules           0.2.8
pytz                     2020.1
requests                 2.23.0
rsa                      4.0
setuptools               46.4.0
six                      1.15.0
uritemplate              3.0.1
urllib3                  1.25.9
wheel                    0.34.2
```

**Comment by plamut** [comment]
@jtressle Thanks for the detailed report!

This sounds a lot like #93, it's a regression caused by one of the recent releases of the `google-api-core` dependency. There is a [fix](https://github.com/googleapis/python-pubsub/issues/93) for it, but i has not yet been merged, nor released.

In the meantime, would it work to downgrade `google-api-core` to `1.16.0`? Please let us know if that solves the problem, thanks!

**Comment by jtressle** [comment]
@plamut thanks for the requirements update.  I was able to test your recommendation, and haven't seen the issue today.  I'll continue to monitor, and will update if I see any further hanging.

Thanks!

**Comment by plamut** [comment]
@jtressle OK great, I'll then close this as a most likely duplicate of #93, but feel free to come back if the issue occurs again.

**Comment by jketcham** [comment]
Huh, so I was just running into this same issue for the past couple hours, trying to find out why all requests to publish to a topic were hanging indefinitely with no exceptions being thrown.

This was only occurring inside a docker container (both locally on my mac and in a GKE cluster); running a python script locally, publishing to the same topic, worked fine.

I was using these versions (the latest as of this writing): 
```
google-api-core==1.26.3 # <
google-cloud-pubsub==2.4.1 # <

# in case these are helpful
google-auth==1.28.0
googleapis-common-protos==1.53.0
grpc-google-iam-v1==0.12.3
grpcio==1.37.0
```

When I downgraded to the versions recommended by @plamut, publishing to the topic started to work:
```
google-api-core==1.16.0 # <
google-cloud-pubsub==1.5.0 # <

# in case these are helpful
google-auth==1.28.0
googleapis-common-protos==1.53.0
grpc-google-iam-v1==0.12.3
grpcio==1.37.0
```

Is there something I'm missing to get this to work correctly with the latest versions?

**Comment by plamut** [comment]
@jketcham Sorry for not seeing this earlier, as this issue is so last year (literally :smile: ).

The issue is about the subscriber (streaming pull, specifically) that stopped working, but you are talking about a _publisher_ problem? Do you use any custom publishing settings, such as customized batching, flow control, and such?

More details would be highly appreciated (ideally as a new issue, because this sounds like a different bug.

Thanks!

**Comment by alejoar** [comment]
I was dealing with this same issue in my Django and Celery instances for a while and downgrading as suggested by @jketcham  didn't help.

Weird thing is that it ONLY failed in Django and Celery. When running the script in it's own stand alone module it would run fine. So I compared the publisher instance inside Django and in the stand alone script (by doing a `print(publisher.__dict__)`) and I noticed something different in the `_batch_lock ` key:

Stand alone:
```
'_batch_lock': <gevent.thread.LockType object at 0x12e5cc5f0>
```

Django/Celery:
```
'_batch_lock': <unlocked _thread.lock object at 0x101b5a7b0>
```

Yes in Django for some reason the threading library was using gevent. That's the problem. Turns out that some time ago I added a [gevent monkey patch](https://www.gevent.org/api/gevent.monkey.html) to my code base and completely forgot about it. THIS WAS THE ISSUE!

I removed all the code dependant of that monkey patch (fortunately just had to rewrite a few lines, the monkey patch was a requirement for a library called grequests) got rid of it.

Now pubsub is working nicely, with the latest version, and I am happy again.
